storm/src/main/storm/mesos/schedulers/DefaultScheduler.java

```diff
@@ -54,6 +55,52 @@ public void prepare(Map conf) {
     mesosStormConf = conf;
   }
 
+  private List<MesosWorkerSlot> getMesosWorkerSlots(Map<String, OfferResources> offerResourcesPerNode,
+                                                    Collection<String> nodesWithExistingSupervisors,
+                                                    TopologyDetails topologyDetails) {
+
+    double requestedWorkerCpu = MesosCommon.topologyWorkerCpu(mesosStormConf, topologyDetails);
+    double requestedWorkerMem = MesosCommon.topologyWorkerMem(mesosStormConf, topologyDetails);
+
+    List<MesosWorkerSlot> mesosWorkerSlots = new ArrayList<>();
+    boolean slotFound = true;
+    int slotsNeeded = topologyDetails.getNumWorkers();
+
+    while (slotFound && slotsNeeded > 0) {
+      slotFound = false;
+      for (String currentNode : offerResourcesPerNode.keySet()) {
+        OfferResources offerResources = offerResourcesPerNode.get(currentNode);
+
+        boolean supervisorExists = nodesWithExistingSupervisors.contains(currentNode);
+
+        if (!offerResources.isFit(mesosStormConf, topologyDetails, supervisorExists)) {
+          log.info("{} is not a fit for {} requestedWorkerCpu: {} requestedWorkerMem: {}",
+                   offerResources.toString(), topologyDetails.getId(), requestedWorkerCpu, requestedWorkerMem);
+          continue;
+        }
+
+        log.info("{} is a fit for {} requestedWorkerCpu: {} requestedWorkerMem: {}", offerResources.toString(),
+                 topologyDetails.getId(), requestedWorkerCpu, requestedWorkerMem);
+        nodesWithExistingSupervisors.add(currentNode);
```

Comment:
We shouldn't add the current node until we have a mesos worker slot for it, so this line should go either in/after the try.